File: firmware/esp32_sensor_node/AnomalyDetector.cpp
#include "AnomalyDetector.h"
#include <math.h>

AnomalyDetector::AnomalyDetector(float matchThreshold)
    : _head(0), _count(0), _consecutiveFails(0), _matchThreshold(matchThreshold)
{
    memset(_window, 0, sizeof(_window));
}

float AnomalyDetector::record(const MatchResult& result, bool success) {
    SignInEvent ev;
    strlcpy(ev.userId, result.userId[0] ? result.userId : "unknown",
            sizeof(ev.userId));
    ev.matchScore  = result.score;
    ev.success     = success;
    ev.timestampMs = millis();
    ev.valid       = true;

    if (success) {
        _consecutiveFails = 0;
    } else {
        if (_consecutiveFails < 255) _consecutiveFails++;
    }

    _window[_head] = ev;
    _head = (_head + 1) % ANOMALY_WINDOW;
    if (_count < ANOMALY_WINDOW) _count++;

    float fail  = _scoreFailureRate();
    float prox  = _scoreMatchProximity(result.score);
    float freq  = _scoreFrequency(ev.timestampMs);
    float score = 0.40f * fail + 0.35f * prox + 0.25f * freq;
    score = score < 0.0f ? 0.0f : (score > 1.0f ? 1.0f : score);

    Serial.printf("[ANOM] score=%.3f  fail=%.2f prox=%.2f freq=%.2f\n",
                  score, fail, prox, freq);
    return score;
}

float AnomalyDetector::bruteForceScore() const {
    return _scoreFailureRate();
}

// BRUTE_FORCE_LIMIT or more consecutive failures → score 1.0
float AnomalyDetector::_scoreFailureRate() const {
    if (_consecutiveFails == 0) return 0.0f;
    float s = (float)_consecutiveFails / BRUTE_FORCE_LIMIT;
    return s > 1.0f ? 1.0f : s;
}

// Match score in the "just barely passed" zone is suspicious
// Score ramps from 0 → 1 as distance approaches the threshold
float AnomalyDetector::_scoreMatchProximity(float score) const {
    float lo = _matchThreshold * 0.30f;  // very confident match — no suspicion
    float hi = _matchThreshold;          // at threshold — max suspicion
    if (score < lo) return 0.0f;
    if (score >= hi) return 0.0f;        // failed match — handled by failure rate
    float s = (score - lo) / (hi - lo);
    return s > 1.0f ? 1.0f : s;
}

// More than 5 events in the last 60 s → anomalous frequency
float AnomalyDetector::_scoreFrequency(uint32_t nowMs) const {
    if (_count == 0) return 0.0f;
    const uint32_t WINDOW_MS = 60000UL;
    const uint8_t  NORMAL_MAX = 5;
    uint8_t recent = 0;
    for (uint8_t i = 0; i < _count; i++) {
        const SignInEvent& e = _window[i];
        if (e.valid && (nowMs - e.timestampMs) < WINDOW_MS) recent++;
    }
    if (recent <= NORMAL_MAX) return 0.0f;
    float s = (float)(recent - NORMAL_MAX) / NORMAL_MAX;
    return s > 1.0f ? 1.0f : s;
}

File: firmware/rpi_sensor_node/anomaly_detector.py
import time
from dataclasses import dataclass
from typing import List

ANOMALY_WINDOW    = 20
BRUTE_FORCE_LIMIT = 5


@dataclass
class SignInEvent:
    user_id: str = ""
    match_score: float = 0.0
    success: bool = False
    timestamp_ms: float = 0.0
    valid: bool = False


class AnomalyDetector:
    def __init__(self, match_threshold: float):
        self._threshold        = match_threshold
        self._window: List[SignInEvent] = [SignInEvent() for _ in range(ANOMALY_WINDOW)]
        self._head             = 0
        self._count            = 0
        self._consecutive_fails = 0

    def record(self, result, success: bool) -> float:
        ev = SignInEvent(
            user_id      = result.user_id if result.user_id else "unknown",
            match_score  = result.score,
            success      = success,
            timestamp_ms = time.time() * 1000,
            valid        = True,
        )

        if success:
            self._consecutive_fails = 0
        else:
            self._consecutive_fails = min(self._consecutive_fails + 1, 255)

        self._window[self._head] = ev
        self._head  = (self._head + 1) % ANOMALY_WINDOW
        if self._count < ANOMALY_WINDOW:
            self._count += 1

        fail  = self._score_failure_rate()
        prox  = self._score_match_proximity(result.score)
        freq  = self._score_frequency(ev.timestamp_ms)
        score = max(0.0, min(1.0, 0.40 * fail + 0.35 * prox + 0.25 * freq))

        print(f"[ANOM] score={score:.3f}  fail={fail:.2f} prox={prox:.2f} freq={freq:.2f}")
        return score

    def brute_force_score(self) -> float:
        return self._score_failure_rate()

    def consecutive_failures(self) -> int:
        return self._consecutive_fails

    # ── Private ───────────────────────────────────────────────────────────────

    def _score_failure_rate(self) -> float:
        if self._consecutive_fails == 0:
            return 0.0
        return min(self._consecutive_fails / BRUTE_FORCE_LIMIT, 1.0)

    def _score_match_proximity(self, score: float) -> float:
        lo = self._threshold * 0.30
        hi = self._threshold
        if score < lo or score >= hi:
            return 0.0
        return min((score - lo) / (hi - lo), 1.0)

    def _score_frequency(self, now_ms: float) -> float:
        if self._count == 0:
            return 0.0
        window_ms  = 60_000.0
        normal_max = 5
        recent = sum(
            1 for ev in self._window
            if ev.valid and (now_ms - ev.timestamp_ms) < window_ms
        )
        if recent <= normal_max:
            return 0.0
        return min((recent - normal_max) / normal_max, 1.0)
